src/cache.py

```diff
@@ -0,0 +1,292 @@
+                "item_id": {"S": "all"},
+                "data": {"S": json.dumps(accounts_data)},
+                "ttl": {"N": str(ttl)},
+                "cached_at": {"S": datetime.now(timezone.utc).isoformat()},
+            },
+        )
+
+        logger.info(f"Cached {len(accounts)} accounts with TTL {cache_config.ttl_minutes} minutes")
+
+    except Exception as e:
+        logger.warning(f"Failed to cache accounts: {e}", extra={"error": str(e)})
+
+
+def get_cached_permission_sets(
+    dynamodb_client: DynamoDBClient,
+    cache_config: CacheConfig,
+    sso_instance_arn: str,
+) -> Optional[list[PermissionSet]]:
+    """Get cached permission sets from DynamoDB.
+
+    Args:
+        dynamodb_client: DynamoDB client
+        cache_config: Cache configuration
+        sso_instance_arn: SSO instance ARN (used as part of cache key)
+
+    Returns:
+        List of cached permission sets or None if cache miss or invalid
+    """
+    if not cache_config.enabled:
+        logger.debug("Cache is disabled, skipping cache lookup")
+        return None
+
+    try:
+        response = dynamodb_client.query(
+            TableName=cache_config.table_name,
+            KeyConditionExpression="cache_key = :cache_key AND item_id = :item_id",
+            ExpressionAttributeValues={
+                ":cache_key": {"S": CacheKey.PERMISSION_SETS},
+                ":item_id": {"S": sso_instance_arn},
+            },
+        )
+
+        items = response.get("Items", [])
+        if not items:
+            logger.info("Cache miss for permission sets")
+            return None
+
+        item = items[0]
+        if not _is_cache_valid(item):
+            logger.info("Cache expired for permission sets")
+            return None
+
+        permission_sets_data = json.loads(item["data"]["S"])
+        permission_sets = [PermissionSet.parse_obj(ps) for ps in permission_sets_data]
+
+        logger.info(f"Cache hit for permission sets, found {len(permission_sets)} permission sets")
+        return permission_sets
+
+    except Exception as e:
+        logger.warning(f"Failed to get cached permission sets: {e}", extra={"error": str(e)})
+        return None
+
+
+def set_cached_permission_sets(
+    dynamodb_client: DynamoDBClient,
+    cache_config: CacheConfig,
+    sso_instance_arn: str,
+    permission_sets: list[PermissionSet],
+) -> None:
+    """Store permission sets in DynamoDB cache.
+
+    Args:
+        dynamodb_client: DynamoDB client
+        cache_config: Cache configuration
+        sso_instance_arn: SSO instance ARN (used as part of cache key)
+        permission_sets: List of permission sets to cache
+    """
+    if not cache_config.enabled:
+        logger.debug("Cache is disabled, skipping cache write")
+        return
+
+    try:
+        ttl = _get_ttl_timestamp(cache_config.ttl_minutes)
+        permission_sets_data = [ps.dict() for ps in permission_sets]
+
+        dynamodb_client.put_item(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Potential NoSQL Injection detected. Untrusted input is being used in a NoSQL database
operation without proper sanitization. NoSQL databases like MongoDB and DynamoDB are not
immune to injection attacks. User input should never be directly interpolated into query
objects or strings. Instead, use parameterized queries, proper sanitization techniques,
or type conversion specific to the NoSQL database you&#39;re using. For MongoDB, consider using
ObjectId() for IDs and validated operators. For DynamoDB, use `boto3.dynamodb.conditions`
classes for safe query construction. Learn more - https://cwe.mitre.org/data/definitions/943.html